### PR TITLE
Re-enable low memory dropout

### PR DESCRIPTION
**Pull request by eellison**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #103330

On attention_is_all_you_need_pytorch:

Perf: 1.526x -> 1.544x      
Memory: 1.00 -> 1.05x

Fix for https://github.com/pytorch/pytorch/issues/102319, although I'm not sure all the perf is recovered.
 

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @ngimel @yf225 @chenyang78 @aakhundov

### CODE REVIEW

File: torch/_inductor/fx_passes/replace_random.py [hunk 1]
```diff
@@ -123,3 +131,62 @@ def replacement(size):
...
+
+
+def _dropout_replacement(x: torch.Tensor, dropout_p: float):
+    assert 0 < dropout_p < 1, "should have been handled in decomps"
+    counters["inductor"]["replace_random"] += 1
+    seed = inductor_prims.seed(x.device)
+    scale = float(1.0 / (1.0 - dropout_p))
+
+    def get_bool_mask():
+        return inductor_prims.random(x.size(), seed, "rand") > dropout_p
```
Comment (ngimel) [inline torch/_inductor/fx_passes/replace_random.py:181]: how would this guarantee that seed is the same between forward and backward, and different between different dropouts/different iterations?

Comment (eellison) [reply torch/_inductor/fx_passes/replace_random.py:181]: Each trace generates a new seed, and we forbid the seed generation from being recomptued - `random_ops += [prims.inductor_seeds, prims.inductor_seed]`.

File: torch/_functorch/partitioners.py [hunk 2]
```diff
@@ -455,6 +457,12 @@ def classify_nodes(joint_module):
         view_ops += [aten.view, aten.slice, aten.permute, aten.t, prims.broadcast_in_dim, aten.expand, aten.as_strided]
         # Natalia said that we should allow recomputing indexing :)
         default_recomputable_ops += [aten.index]
+
+        # this takes in a seed, so it's deterministic, and we can recompute it
+        default_recomputable_ops += [prims.inductor_random]
+
+        random_ops += [prims.inductor_seeds, prims.inductor_seed]
```
Comment (anijain2305) [inline torch/_functorch/partitioners.py:464]: We should probably recompute seeds? This might change rng behavior.

Comment (eellison) [reply torch/_functorch/partitioners.py:464]: This does prevent recomputation of seeds

Comment (anijain2305) [reply torch/_functorch/partitioners.py:464]: Ah, sorry, misssed it.

**Review by jansel — APPROVED** [review]

**Comment by malfet** [comment]
See https://hud.pytorch.org/pytorch/pytorch/commit/2d14395f176b38b8416c2713d285e5ae55695a5f

**Comment by davidberard98** [comment]
@eellison This caused 117s -> 149s compilation latency on dynamic huggingface [link](https://hud.pytorch.org/benchmark/compilers?startTime=Fri%2C%2023%20Jun%202023%2022%3A35%3A39%20GMT&stopTime=Fri%2C%2030%20Jun%202023%2022%3A35%3A39%20GMT&granularity=hour&suite=torchbench&mode=training&dtype=amp&lBranch=main&lCommit=7bc181d374a54db4b1556d263be965f6a848eecb&rBranch=main&rCommit=d4e51511a0e471072359800b03e1b63f2ef2b3f6), e.g. 110s -> 386s T5Small

**Comment by eellison** [comment]
Wow, that is a large regression for T5Small. I'll revert until we can figure that out further.

**Comment by Chillee** [comment]
I'd guess it probably has something to do with preventing each dropout kernel from reusing the same kernel?
